File: sdks/sandbox/javascript/src/factory/defaultAdapterFactory.ts
// Copyright 2026 Alibaba Group Holding Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createExecdClient } from "../openapi/execdClient.js";
import { createEgressClient } from "../openapi/egressClient.js";
import { createLifecycleClient } from "../openapi/lifecycleClient.js";

import { CommandsAdapter } from "../adapters/commandsAdapter.js";
import { EgressAdapter } from "../adapters/egressAdapter.js";
import { FilesystemAdapter } from "../adapters/filesystemAdapter.js";
import { HealthAdapter } from "../adapters/healthAdapter.js";
import { MetricsAdapter } from "../adapters/metricsAdapter.js";
import { SandboxesAdapter } from "../adapters/sandboxesAdapter.js";

import type {
  AdapterFactory,
  CreateEgressStackOptions,
  CreateExecdStackOptions,
  CreateLifecycleStackOptions,
  EgressStack,
  ExecdStack,
  LifecycleStack,
} from "./adapterFactory.js";

export class DefaultAdapterFactory implements AdapterFactory {
  createLifecycleStack(opts: CreateLifecycleStackOptions): LifecycleStack {
    const lifecycleClient = createLifecycleClient({
      baseUrl: opts.lifecycleBaseUrl,
      apiKey: opts.connectionConfig.apiKey,
      headers: opts.connectionConfig.headers,
      fetch: opts.connectionConfig.fetch,
    });
    const sandboxes = new SandboxesAdapter(lifecycleClient);
    return { sandboxes };
  }

  createExecdStack(opts: CreateExecdStackOptions): ExecdStack {
    const headers: Record<string, string> = {
      ...(opts.connectionConfig.headers ?? {}),
      ...(opts.endpointHeaders ?? {}),
    };
    const execdClient = createExecdClient({
      baseUrl: opts.execdBaseUrl,
      headers,
      fetch: opts.connectionConfig.fetch,
    });

    const health = new HealthAdapter(execdClient);
    const metrics = new MetricsAdapter(execdClient);
    const files = new FilesystemAdapter(execdClient, {
      baseUrl: opts.execdBaseUrl,
      fetch: opts.connectionConfig.fetch,
      headers,
    });
    const commands = new CommandsAdapter(execdClient, {
      baseUrl: opts.execdBaseUrl,
      fetch: opts.connectionConfig.sseFetch,
      headers,
    });

    return {
      commands,
      files,
      health,
      metrics,
    };
  }

  createEgressStack(opts: CreateEgressStackOptions): EgressStack {
    const headers: Record<string, string> = {
      ...(opts.connectionConfig.headers ?? {}),
      ...(opts.endpointHeaders ?? {}),
    };
    const egressClient = createEgressClient({
      baseUrl: opts.egressBaseUrl,
      headers,
      fetch: opts.connectionConfig.fetch,
    });
    const egress = new EgressAdapter(egressClient, {
      baseUrl: opts.egressBaseUrl,
      fetch: opts.connectionConfig.fetch,
      headers,
    });
    return {
      egress,
      credentialVault: egress,
    };
  }
}

export function createDefaultAdapterFactory(): AdapterFactory {
  return new DefaultAdapterFactory();
}

File: sdks/sandbox/csharp/src/OpenSandbox/Factory/DefaultAdapterFactory.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using OpenSandbox.Adapters;
using OpenSandbox.Internal;
using Microsoft.Extensions.Logging;

namespace OpenSandbox.Factory;

/// <summary>
/// Default implementation of the adapter factory.
/// </summary>
public sealed class DefaultAdapterFactory : IAdapterFactory
{
    /// <summary>
    /// Creates a new instance of the default adapter factory.
    /// </summary>
    /// <returns>A new adapter factory instance.</returns>
    public static IAdapterFactory Create() => new DefaultAdapterFactory();

    /// <inheritdoc />
    public LifecycleStack CreateLifecycleStack(CreateLifecycleStackOptions options)
    {
        var clientWrapper = new HttpClientWrapper(
            options.HttpClientProvider.HttpClient,
            options.LifecycleBaseUrl,
            options.ConnectionConfig.Headers,
            options.LoggerFactory.CreateLogger("OpenSandbox.HttpClientWrapper"));

        var sandboxes = new SandboxesAdapter(clientWrapper);

        return new LifecycleStack
        {
            Sandboxes = sandboxes
        };
    }

    /// <inheritdoc />
    public ExecdStack CreateExecdStack(CreateExecdStackOptions options)
    {
        var headers = options.ExecdHeaders ?? options.ConnectionConfig.Headers;

        var clientWrapper = new HttpClientWrapper(
            options.HttpClientProvider.HttpClient,
            options.ExecdBaseUrl,
            headers,
            options.LoggerFactory.CreateLogger("OpenSandbox.HttpClientWrapper"));

        var health = new HealthAdapter(clientWrapper);
        var metrics = new MetricsAdapter(clientWrapper);
        var files = new FilesystemAdapter(
            clientWrapper,
            options.HttpClientProvider.HttpClient,
            options.ExecdBaseUrl,
            headers);
        var commands = new CommandsAdapter(
            clientWrapper,
            options.HttpClientProvider.SseHttpClient,
            options.ExecdBaseUrl,
            headers,
            options.LoggerFactory.CreateLogger("OpenSandbox.CommandsAdapter"));

        return new ExecdStack
        {
            Commands = commands,
            Files = files,
            Health = health,
            Metrics = metrics
        };
    }

    /// <inheritdoc />
    public EgressStack CreateEgressStack(CreateEgressStackOptions options)
    {
        var headers = options.EgressHeaders ?? options.ConnectionConfig.Headers;

        var clientWrapper = new HttpClientWrapper(
            options.HttpClientProvider.HttpClient,
            options.EgressBaseUrl,
            headers,
            options.LoggerFactory.CreateLogger("OpenSandbox.HttpClientWrapper"));

        var egress = new EgressAdapter(clientWrapper);
        return new EgressStack
        {
            Egress = egress,
            CredentialVault = egress
        };
    }
}
